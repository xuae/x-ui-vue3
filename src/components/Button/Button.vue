<template>
  <button :type="nativeType" :class="classes">Button</button>
</template>

<script lang="ts">
import { computed, defineComponent, Prop, toRefs } from 'vue';

export default defineComponent({
  name: 'XButton',
  props: {
    // 原生 type 属性
    nativeType: {
      type: String,
      validator: (val: string) => {
        return ['button', 'submit', 'reset'].includes(val);
      },
      default: 'button',
    } as Prop<'button' | 'submit' | 'reset'>,

    // 样式
    stylus: {
      type: String,
      validator: (val: string) => {
        return ['text', 'icon', 'outlined', 'contained', 'toggle'].includes(
          val
        );
      },
    } as Prop<'text' | 'icon' | 'outlined' | 'contained' | 'toggle'>,

    // 类型
    type: {
      type: String,
      validator: (val: string) => {
        return ['flat', 'fab'].includes(val);
      },
    } as Prop<'flat' | 'fab'>,

    // 点击是否有波纹效果
    ripple: {
      type: Boolean,
      default: true,
    },

    // 圆角按钮
    round: {
      type: Boolean,
      default: false,
    },

    // 小尺寸按钮
    small: {
      type: Boolean,
      default: false,
    },

    // 块状按钮
    block: {
      type: Boolean,
      default: false,
    },

    // 禁用按钮
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const { stylus, type, ripple, round, small, block, disabled } = toRefs(
      props
    );
    const classes = computed(() => [
      'x-button',
      stylus ? `x-button__${stylus.value}` : undefined,
      type ? `x-button__${type.value}` : undefined,
      {
        'x-button__ripple': ripple.value,
        'x-button__round': round.value,
        'x-button__small': small.value,
        'x-button__block': block.value,
        'x-button__disabled': disabled.value,
      },
    ]);

    return {
      classes,
    };
  },
});
</script>

<style lang="scss" scoped>
@import 'Button';
</style>
