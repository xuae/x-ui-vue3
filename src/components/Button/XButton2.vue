<template>
  <button :type="nativeType" :class="classes">Button2</button>
</template>

<script>
import { computed, defineComponent, toRefs } from 'vue';

const nativeTypes = ['button', 'submit', 'reset'];

const styluses = ['text', 'icon', 'outlined', 'contained', 'toggle'];

const types = ['flat', 'fab'];

export default defineComponent({
  name: 'XButton2',
  props: {
    // button 原生 type
    nativeType: {
      type: String,
      validator(val) {
        return nativeTypes.includes(val);
      },
    },

    // 样式
    stylus: {
      type: String,
      validator(val) {
        return styluses.includes(val);
      },
    },

    // 类型
    type: {
      type: String,
    },

    // 点击是否有波纹效果
    ripple: {
      type: Boolean,
      default: true,
    },

    // 是否圆角按钮
    round: {
      type: Boolean,
      default: false,
    },

    // 是否小尺寸按钮
    small: {
      type: Boolean,
      default: false,
    },

    // 是否是块状按钮，填充整行
    block: {
      type: Boolean,
      default: false,
    },

    // 是否禁用按钮
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const { stylus, type, small, ripple, round, disabled } = toRefs(props);
    const classes = computed(() => [
      'x-button',
      stylus ? `x-button__${stylus.value}` : '',
      type ? `x-button__${type.value}` : '',
      {
        'x-button__small': small.value,
        'x-button__ripple': ripple.value,
        'x-button__round': round.value,
        'x-button__disabled': disabled.value,
      },
    ]);

    return {
      classes,
    };
  },
});
</script>

<style lang="scss" scoped>
@import 'XButton';
</style>
