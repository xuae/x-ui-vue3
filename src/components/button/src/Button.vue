<template>
  <button :type="nativeType" :class="classes" v-bind="$attrs">
    <slot />
  </button>
</template>

<script lang="ts">
import { computed, defineComponent, Prop, toRefs } from 'vue';
import Util from '@/utils/Util';
import VueProp from '@/utils/VueProp';

export default defineComponent({
  name: 'XButton',
  props: {
    // 块状按钮
    block: {
      type: Boolean,
      default: false,
    },

    // 颜色
    color: {
      type: String,
      validator: VueProp.oneOf([
        'primary',
        'secondary',
        'info',
        'success',
        'warning',
        'error',
      ]),
    } as Prop<
      'primary' | 'secondary' | 'info' | 'success' | 'warning' | 'error'
    >,

    // 禁用按钮
    disabled: {
      type: Boolean,
      default: false,
    },

    // 原生 type 属性
    nativeType: {
      type: String,
      validator: VueProp.oneOf(['button', 'submit', 'reset']),
      default: 'button',
    } as Prop<'button' | 'submit' | 'reset'>,

    // 点击是否有波纹效果
    ripple: {
      type: Boolean,
      default: true,
    },

    // 圆角按钮
    round: {
      type: Boolean,
      default: false,
    },

    // 小尺寸按钮
    small: {
      type: Boolean,
      default: false,
    },

    // 样式
    stylus: {
      type: String,
      validator: VueProp.oneOf([
        'text',
        'icon',
        'outlined',
        'contained',
        'toggle',
      ]),
    } as Prop<'text' | 'icon' | 'outlined' | 'contained' | 'toggle'>,

    // 类型
    type: {
      type: String,
      validator: VueProp.oneOf(['flat', 'fab']),
    } as Prop<'flat' | 'fab'>,
  },
  setup(props) {
    const {
      block,
      color,
      disabled,
      ripple,
      round,
      small,
      stylus,
      type,
    } = toRefs(props);
    const classes = computed(() =>
      Util._.compact([
        'x-button',
        color ? `x-button__${color.value}` : undefined,
        stylus ? `x-button__${stylus.value}` : undefined,
        type ? `x-button__${type.value}` : undefined,
        {
          'x-button__block': block.value,
          'x-button__disabled': disabled.value,
          'x-button__ripple': ripple.value,
          'x-button__round': round.value,
          'x-button__small': small.value,
        },
      ])
    );

    return {
      classes,
    };
  },
});
</script>
