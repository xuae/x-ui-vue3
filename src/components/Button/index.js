// 配置对外引用
import './Button.scss';
import Button from './Button.vue';

// 提供 install 方法
// 这里提供一次 install 是为了便于单独引入组件时进行注册
Button.install = function (Vue) {
  Vue.component(Button.name, Button);
};

// 默认导出方式导出
export default Button;
