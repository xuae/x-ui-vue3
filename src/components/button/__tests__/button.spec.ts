import { mount } from '@vue/test-utils';
import Button from '@/components/button/src/Button.vue';

describe('Button.vue', () => {
  it('create', () => {
    const wrapper = mount(Button);
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.classes('x-button')).toBe(true);
  });

  it('slot default', () => {
    const text = '按钮';
    const wrapper = mount(Button, {
      slots: {
        default: text,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.text()).toContain(text);
  });

  it('prop nativeType', async () => {
    const data = ['button', 'submit', 'reset'];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          nativeType: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.attributes('type')).toBe(data[i]);
    }
  });

  it('prop stylus', async () => {
    const data = ['text', 'icon', 'outlined', 'contained', 'toggle'];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          stylus: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes(`x-button__${data[i]}`)).toBe(true);
    }
  });

  it('prop type', () => {
    const data = ['flat', 'fab'];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          type: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes(`x-button__${data[i]}`)).toBe(true);
    }
  });

  it('prop ripple', () => {
    const data = [true, false];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          ripple: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes('x-button__ripple')).toBe(data[i]);
    }
  });

  it('prop round', () => {
    const data = [true, false];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          round: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes('x-button__round')).toBe(data[i]);
    }
  });

  it('prop small', () => {
    const data = [true, false];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          small: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes('x-button__small')).toBe(data[i]);
    }
  });

  it('prop block', async () => {
    const data = [true, false];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          block: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes('x-button__block')).toBe(data[i]);
    }
  });

  it('prop disabled', () => {
    const data = [true, false];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Button as any, {
        props: {
          disabled: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes('x-button__disabled')).toBe(data[i]);
    }
  });
});
