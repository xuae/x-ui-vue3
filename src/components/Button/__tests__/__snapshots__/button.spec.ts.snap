// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue block 1`] = `
<button
  class="x-button   x-button__ripple x-button__block"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue block 2`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue create 1`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue disabled 1`] = `
<button
  class="x-button   x-button__ripple x-button__disabled"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue disabled 2`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue nativeType 1`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue nativeType 2`] = `
<button
  class="x-button   x-button__ripple"
  type="submit"
>
  Button
</button>
`;

exports[`Button.vue nativeType 3`] = `
<button
  class="x-button   x-button__ripple"
  type="reset"
>
  Button
</button>
`;

exports[`Button.vue ripple 1`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue ripple 2`] = `
<button
  class="x-button"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue round 1`] = `
<button
  class="x-button   x-button__ripple x-button__round"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue round 2`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue small 1`] = `
<button
  class="x-button   x-button__ripple x-button__small"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue small 2`] = `
<button
  class="x-button   x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue stylus 1`] = `
<button
  class="x-button x-button__text  x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue stylus 2`] = `
<button
  class="x-button x-button__icon  x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue stylus 3`] = `
<button
  class="x-button x-button__outlined  x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue stylus 4`] = `
<button
  class="x-button x-button__contained  x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue stylus 5`] = `
<button
  class="x-button x-button__toggle  x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue type 1`] = `
<button
  class="x-button  x-button__flat x-button__ripple"
  type="button"
>
  Button
</button>
`;

exports[`Button.vue type 2`] = `
<button
  class="x-button  x-button__fab x-button__ripple"
  type="button"
>
  Button
</button>
`;
